### DISCUSSION TITLE
Key combination / Key chords

### DISCUSSION THREAD

**Discussion by egolep** [YaLTeR/niri#416] [Ideas]
Many WM (I know about Sway/i3 and Qtile, but I'm pretty sure there are others) offers the possibility to create "key chord", i.e. there is a first combination of keys that, when pressed, triggers a sub-set of key that can be pressed in order to then do something.
In Sway, for example, I have a main shortcut wich is Mod+d that I call to my self the "launch programs" combination, and then I can press one of different keys to launch a specific application (like "b" for a browser). This helps me organizing my keybindings and has helped me having a huge amount of very specific keybindings for many different common operations without the risk of having conflicts.
How hard would it be to implement something like this in Niri? And could it be interesting for the project or is it just a spoil of mine?

**Comment by YaLTeR** [comment]
Maybe not too difficult this way, but I'm not entirely sure if we want that.

**Reply by egolep** [reply]
You mean for code maintainability/quality or because you think it's not a useful feature?

**Reply by YaLTeR** [reply]
These two form a balance, don't they :) It's extra complexity, both for code and for configuration, that I'm not sure is warranted for how much usefulness it brings.

**Reply by egolep** [reply]
If you mean user configuration, to me it's more like a "don't want it / don't use it" kind of thing. But it sure does add complexity to development and maintenance, so it's really fair to not do this if it's not something that the community can be interested in.

Thanks for your response!

**Reply by RayNguyen842857** [reply]
It might just be me, but I have a rather valid use case for this feature beyond simply triggering different key bindings (though I do find that aspect very useful, especially as a polyglot like me who needs to switch between various language keyboard layouts, though this I could manage without).

Currently, I'm utilizing the submap feature, which is the same feature available in Hyprland. I've set my right Alt key to a completely blank subset, effectively disabling all other key bindings until I press right Alt again to exit that mode. This is extremely useful when I need to connect to my Windows gaming VM or my company's VDI as with all key bindings temporarily disabled by going inside that mode, I can fully utilize the VM/VDI shortcuts without any interference.

**Reply by YaLTeR** [reply]
> This is extremely useful when I need to connect to my Windows gaming VM or my company's VDI as with all key bindings temporarily disabled by going inside that mode, I can fully utilize the VM/VDI shortcuts without any interference.

There's a proper solution for this: #198

**Comment by YaLTeR** [comment]
There's a WIP implementation for binding modes: #532 The author's GitHub account had disappeared, but I remember about the PR and will get to it at some point. I still have some questions remaining like: what's a good way to decide whether a binding mode exits to normal on any key press, or keeps enabled until explicitly disabled?

**Reply by RayNguyen842857** [reply]
I believe the default behavior in Hyprland is to keep the mode enabled until an explicit signal is given to exit. That way we can always chain action such as executing something && exit mode to return to normal mode again or else you don't have to do that if you want to stay in that mode even after executing something (such as in resizing window).
